server/tests/integrations/stripe/test_payment_dunning.py

```diff
@@ -0,0 +1,106 @@
+import uuid
+from datetime import timedelta
+
+import pytest
+import stripe as stripe_lib
+from freezegun import freeze_time
+from pytest_mock import MockerFixture
+
+from polar.integrations.stripe.payment import handle_failure
+from polar.kit.db.postgres import AsyncSession
+from polar.kit.utils import utc_now
+from polar.models import (
+    Customer,
+    Product,
+    Subscription,
+)
+from polar.models.subscription import SubscriptionStatus
+from polar.order.repository import OrderRepository
+from polar.subscription.repository import SubscriptionRepository
+from tests.fixtures.database import SaveFixture
+from tests.fixtures.random_objects import (
+    create_order,
+)
+
+
+@pytest.fixture
+def stripe_charge() -> stripe_lib.Charge:
+    return stripe_lib.Charge(
+        id="ch_test_123",
+        object="charge",
+        amount=2000,
+        currency="usd",
+        status="failed",
+        metadata={"checkout_id": str(uuid.uuid4())},
```

Comment:
To resolve the order from the charge, you should set metadata like this:

```py
metadata={"order_id": str(order.id)},
```

To have the order at hand, you have two possibilities:

1. Don't create a separate fixture for `stripe_charge`, simply declare it in the test. Probably the easiest since we only have one test right now
2. Declare an order fixture first and inject it in this fixture.

---

We can discard the `checkout_id` metadata, since we're not testing the checkout resolution here (i.e. `resolve_checkout` will return `None`).